File: serve.py
#!/usr/bin/env python3
"""Servidor estático de Wardern SIN caché (para que los cambios aparezcan al refrescar)."""
import http.server
import socketserver
import sys

PORT = 8099
if "--port" in sys.argv:
    PORT = int(sys.argv[sys.argv.index("--port") + 1])


class NoCacheHandler(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Cache-Control", "no-cache, no-store, must-revalidate")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()


class TCPServer(socketserver.TCPServer):
    allow_reuse_address = True


if __name__ == "__main__":
    with TCPServer(("0.0.0.0", PORT), NoCacheHandler) as httpd:
        print(f"Wardern en http://localhost:{PORT} (sin caché)")
        httpd.serve_forever()

File: tools/serve.mjs
#!/usr/bin/env node
// Servidor estático de Wardern SIN caché, equivalente en Node de serve.py.
// Existe porque no todas las máquinas de desarrollo tienen Python instalado.
//   node tools/serve.mjs [--port 8099]
import http from "node:http";
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const i = process.argv.indexOf("--port");
const PORT = i === -1 ? 8099 : Number(process.argv[i + 1]);

const TIPOS = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".webp": "image/webp",
  ".ico": "image/x-icon",
  ".woff2": "font/woff2",
};

const servidor = http.createServer(async (req, res) => {
  const sinQuery = decodeURIComponent(req.url.split("?")[0]);
  // Resolver dentro de ROOT: bloquea ../ y rutas absolutas.
  const destino = path.join(ROOT, path.normalize(sinQuery).replace(/^(\.\.[/\\])+/, ""));
  if (!destino.startsWith(ROOT)) {
    res.writeHead(403).end("403");
    return;
  }

  let archivo = destino;
  try {
    if ((await stat(archivo)).isDirectory()) archivo = path.join(archivo, "index.html");
    await stat(archivo);
  } catch {
    res.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" }).end("404");
    return;
  }

  res.writeHead(200, {
    "Content-Type": TIPOS[path.extname(archivo).toLowerCase()] ?? "application/octet-stream",
    "Cache-Control": "no-cache, no-store, must-revalidate",
    Pragma: "no-cache",
    Expires: "0",
  });
  createReadStream(archivo).pipe(res);
});

servidor.listen(PORT, () => console.log(`Wardern en http://localhost:${PORT}`));
